[SPARK-6113] [ml] Tree ensembles for Pipelines API

jkbradley opened the pull request:
This is a continuation of [https://github.com/apache/spark/pull/5530](which was for Decision Trees), but for ensembles: Random Forests and Gradient-Boosted Trees.  Please refer to the JIRA [https://issues.apache.org/jira/browse/SPARK-6113], the design doc linked from the JIRA, and the previous PR linked above for design discussions.

This PR follows the example set by the previous PR for Decision Trees.  It includes a few cleanups to Decision Trees.

Note: There is one issue which will be addressed in a separate PR: Ensembles' component Models have no parent or fittingParamMap.  I plan to submit a separate PR which makes those values in Model be Options.  It does not matter much which PR gets merged first.

CC: @mengxr @manishamde @codedeft @chouqin

jkbradley committed:
Added files from old PR and did some initial updates.

jkbradley committed:
some more updates, but far from done

jkbradley committed:
Added random forests and test suites to spark.ml.  Not tested yet.  Need to add example as well

jkbradley committed:
Added GBT to spark.ml, with tests and examples

SparkQA commented:
[Test build #30729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30729/consoleFull) for   PR 5626 at commit [`ea3d901`](https://github.com/apache/spark/commit/ea3d901a65ac3d1a2e93ecf089dfbd3c7a91cae9).

SparkQA commented:
[Test build #30729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30729/consoleFull) for   PR 5626 at commit [`ea3d901`](https://github.com/apache/spark/commit/ea3d901a65ac3d1a2e93ecf089dfbd3c7a91cae9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Params(`
  * `  case class Params(`
  * `final class GBTClassificationModel(`
  * `final class GBTRegressionModel(`
  * `trait TreeEnsembleModel `
  * `case class Explode(child: Expression)`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30729/
Test FAILed.

jkbradley committed:
scala style fix

SparkQA commented:
[Test build #30764 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30764/consoleFull) for   PR 5626 at commit [`855aa9a`](https://github.com/apache/spark/commit/855aa9aeb087576c3ffeed7f27ec5e18c23ac67f).

SparkQA commented:
[Test build #30764 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30764/consoleFull) for   PR 5626 at commit [`855aa9a`](https://github.com/apache/spark/commit/855aa9aeb087576c3ffeed7f27ec5e18c23ac67f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Params(`
  * `  case class Params(`
  * `final class GBTClassificationModel(`
  * `final class GBTRegressionModel(`
  * `trait TreeEnsembleModel `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30764/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala:
```diff
@@ -0,0 +1,225 @@
+
+  override def setCacheNodeIds(value: Boolean): this.type = super.setCacheNodeIds(value)
+
+  override def setCheckpointInterval(value: Int): this.type = super.setCheckpointInterval(value)
```
Should it go to shared params? I see the problem with the doc. If we want to put something special, we can put it in the JavaDoc. No strong preference about this. But it makes me think that whether we should mark shared params final.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala:
```diff
@@ -0,0 +1,225 @@
+
+  /**
+   * Loss function which GBT tries to minimize. (case-insensitive)
+   * Supported: "LogLoss"
```
Though the values are case insensitive, they should show up consistently in the doc. `logloss` or `logLoss`? The latter looks better to me. Btw, is `log` sufficient?

jkbradley replied:
How about:
* classification: "logistic"
* regression: "absolute" and "squared" (or "L1" and "L2"?)

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala:
```diff
@@ -0,0 +1,225 @@
+   * (default = LogLoss)
+   * @group param
+   */
+  val loss: Param[String] = new Param[String](this, "loss", "Loss function which GBT tries to" +
```
`loss` -> `lossType`? `loss` may be too general. If `loss` appears in another algorithm, they should have similar semantic. For example, if we put `loss` as an evaluation metric, it becomes a numerical value. No strong preference.

jkbradley replied:
I'll go for lossType

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala:
```diff
@@ -0,0 +1,225 @@
+
+object GBTClassifier {
+  // The losses below should be lowercase.
+  /** Accessor for supported loss settings */
```
Should list available options in the JavaDoc.

jkbradley replied:
They are in the Param doc.  Should they be here too?

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala:
```diff
@@ -0,0 +1,225 @@
+final class GBTClassificationModel(
+    override val parent: GBTClassifier,
+    override val fittingParamMap: ParamMap,
+    val trees: Array[DecisionTreeRegressionModel],
```
If we want to support multiclass in the future, is the type `Array[DecisionTreeRegressionModel]` still compatible? Btw, `trees` and `treeWeights` are public. See my comment on `TreeEnsembleModel`.

jkbradley replied:
Sort of.  Gradient boosting's "gradient" part comes from phrasing the problem as regression.

scikit-learn trains 1 regression tree for each class, essentially doing a 1 vs all approach to reduce multiclass to binary.  Those predictions can then be fed into the boosting algorithm via a multinomial logistic loss function.

Just as in reducing multiclass to binary, we can take other approaches.  E.g., if you use an error-correcting code-based approach, then it can make sense to train 1 tree per iteration, where each tree tries to separate some subset of classes out from the other classes.

I have seen (a) scikit-learn's approach of 1-vs-all for gradient boosting and (b) the error-correcting approach for other boosting algorithms, but I have not seen someone write out the combination: gradient boosting + error-correcting.  It may be out there, and I'm sure we could work it out, with a bit of effort.

So, this abstraction will work if we take the error-correcting approach.  It could work for 1-vs-all if we keep a matching array indicating which class each tree is testing for, but then the ordering of the trees would be obscured.  For 1-vs-all, it might be best to store ```Array[Array[DecisionTreeRegressionModel]]```

mengxr replied:
Got it. Thanks!

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala:
```diff
@@ -0,0 +1,180 @@
+}
+
+object RandomForestClassifier {
+  /** Accessor for supported impurity settings */
```
Need to put available options in the JavaDoc.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/impl/tree/treeParams.scala:
```diff
@@ -298,3 +302,200 @@ private[ml] object TreeRegressorParams {
+   * Random seed for bootstrapping and choosing feature subsets.
+   * @group param
+   */
+  final val seed: LongParam = new LongParam(this, "seed",
```
This should go to shared params.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/impl/tree/treeParams.scala:
```diff
@@ -298,3 +302,200 @@ private[ml] object TreeRegressorParams {
+  }
+
+  /** @group getParam */
+  def getSubsamplingRate: Double = getOrDefault(subsamplingRate)
```
Most getters should be final.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/impl/tree/treeParams.scala:
```diff
@@ -298,3 +302,200 @@ private[ml] object TreeRegressorParams {
+   * @group param
+   */
+  final val featuresPerNode: Param[String] = new Param[String](this, "featuresPerNode",
+    "The number of features to consider for splits at each tree node." +
```
`The number of features` -> `Strategy to choose a subset of features for splits.` Is the original name `featureSubsetStrategy` more accurate?

jkbradley replied:
"Strategy" makes it sound like it's something fancier than just setting a number based on the total number of features.  I'm ambivalent, though; there aren't any great options.

jkbradley replied:
I"ll switch back to featureSubsetStrategy since users may be used to it.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/impl/tree/treeParams.scala:
```diff
@@ -298,3 +302,200 @@ private[ml] object TreeRegressorParams {
+   * (default = 0.1)
+   * @group param
+   */
+  final val learningRate: DoubleParam = new DoubleParam(this, "learningRate",
```
Unrelated question: When should we use `learningRate` instead of `stepSize` in general?

jkbradley replied:
Oops, I meant to change that, per a discussion from another PR.  I forget where now...but my vote was for stepSize.  I think they are synonymous, so I could put one (stepSize?) in sharedParams.  What do you think?

jkbradley replied:
I'm hesitating about putting it in sharedParams since the intended range can differ between algorithms.  For GBTs, it should be in (0, 1], but it could be different for other algs.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:
```diff
@@ -58,3 +58,43 @@ trait DecisionTreeModel {
+ * TODO: Add support for predicting probabilities and raw predictions
+ */
+@AlphaComponent
+trait TreeEnsembleModel {
```
Should it be public? Note that adding method to an interface counts as a break change.

jkbradley replied:
Yeah, I think I opened it up in order to write examples which worked for both classification and regression.  I'll change those.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:
```diff
@@ -58,3 +58,43 @@ trait DecisionTreeModel {
+  //       DecisionTreeModel.
+
+  /** Trees in this ensemble. Warning: These have null parent Estimators. */
+  def getTrees: Array[DecisionTreeModel]
```
If we don't allow users to set trees, it is a property and maybe calling it `trees` would be sufficient. Same for `getTreeWeights`.

mengxr commented:
@jkbradley I made one pass on the public APIs. There are some issues from the ml.DT PR:

1. Node.prediction should say "leaf" node instead of "internal":
https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala#L33
2. CategoricalSplit.getLeftCategories. Maybe this should be a property since it is immutable. No strong preference. Btw, I'm thinking about using `BitSet` to store left category indices to save storage.

We can address those in a separate PR. I will take another pass on the implementation.

jkbradley committed:
Updated per all comments in code review

jkbradley commented:
Updated!  I think the only thing I didn't do was make stepSize a shared param.  Copying from the comment above:
> I'm hesitating about putting it in sharedParams since the intended range can differ between algorithms. For GBTs, it should be in (0, 1], but it could be different for other algs.

I updated the doc in Node.prediction, as well as getLeft/RightCategories.  I'll make a JIRA for using BitSet internally for categories.

SparkQA commented:
[Test build #30882 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30882/consoleFull) for   PR 5626 at commit [`bbae2a2`](https://github.com/apache/spark/commit/bbae2a2a98b6df9ef0f4d32532e52943ddd15694).

SparkQA commented:
[Test build #30882 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30882/consoleFull) for   PR 5626 at commit [`bbae2a2`](https://github.com/apache/spark/commit/bbae2a2a98b6df9ef0f4d32532e52943ddd15694).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Params(`
  * `  case class Params(`
  * `final class GBTClassificationModel(`
  * `trait HasSeed extends Params `
  * `final class GBTRegressionModel(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30882/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala:
```diff
@@ -0,0 +1,226 @@
+    val numClasses: Int = MetadataUtils.getNumClasses(dataset.schema(paramMap(labelCol))) match {
+      case Some(n: Int) => n
+      case None => throw new IllegalArgumentException("GBTClassifier was given input" +
+        s" with invalid label column, without the number of classes specified.")
```
Mention `StringIndexer`? Btw, we should pair TODOs with JIRAs.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala:
```diff
@@ -0,0 +1,184 @@
+    val numClasses: Int = MetadataUtils.getNumClasses(dataset.schema(paramMap(labelCol))) match {
+      case Some(n: Int) => n
+      case None => throw new IllegalArgumentException("RandomForestClassifier was given input" +
+        s" with invalid label column, without the number of classes specified.")
```
Mention the label column name in the error message.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala:
```diff
@@ -0,0 +1,184 @@
+  override def trees: Array[DecisionTreeModel] = _trees.asInstanceOf[Array[DecisionTreeModel]]
+
+  // Note: We may add support for weights (based on tree performance) later on.
+  lazy val _treeWeights: Array[Double] = Array.fill[Double](numTrees)(1.0)
```
private

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/impl/tree/treeParams.scala:
```diff
@@ -296,5 +299,194 @@ private[ml] trait TreeRegressorParams extends Params {
+    require(value > 0.0 && value <= 1.0,
+      s"Subsampling rate must be in range (0,1]. Bad rate: $value")
+    set(subsamplingRate, value)
+    this
```
`this` is not required.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/impl/tree/treeParams.scala:
```diff
@@ -296,5 +299,194 @@ private[ml] trait TreeRegressorParams extends Params {
+  final def getSubsamplingRate: Double = getOrDefault(subsamplingRate)
+
+  /** @group setParam */
+  def setSeed(value: Long): this.type = {
```
`= set(seed.value)` should be sufficient.

mengxr commented on mllib/src/test/java/org/apache/spark/ml/classification/JavaGBTClassifierSuite.java:
```diff
@@ -0,0 +1,100 @@
+    double B = -1.5;
+
+    JavaRDD<LabeledPoint> data = sc.parallelize(
+        LogisticRegressionSuite.generateLogisticInputAsList(A, B, nPoints, 42), 2).cache();
```
2-space indentation

mengxr commented on mllib/src/test/java/org/apache/spark/ml/classification/JavaGBTClassifierSuite.java:
```diff
@@ -0,0 +1,100 @@
+        .setMaxIter(3)
+        .setStepSize(0.1)
+        .setMaxDepth(2); // duplicate setMaxDepth to check builder pattern
+    for (int i = 0; i < GBTClassifier.supportedLossTypes().length; ++i) {
```
~~~java
for (String lossType: GBTClassifier.supportedLossTypes()) {
 ...
}
~~~

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  test("alternating categorical and continuous features with multiclass labels to test indexing") {
+    val arr = new Array[LabeledPoint](4)
```
~~~scala
val arr = Array(
  LabeledPoint(0.0, Vectors.dense(1.0, 0.0, 0.0, 3.0, 1.0)),
  ...)
~~~

mengxr commented on examples/src/main/scala/org/apache/spark/examples/ml/GBTExample.scala:
```diff
@@ -0,0 +1,238 @@
+        .required()
+        .action((x, c) => c.copy(input = x))
+      checkConfig { params =>
+        if (params.fracTest < 0 || params.fracTest > 1) {
```
`>= 1` or document the behavior when `fracTest  == 1`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/DecisionTreeClassifier.scala:
```diff
@@ -85,18 +82,16 @@ final class DecisionTreeClassifier
   }
 
   /** (private[ml]) Create a Strategy instance to use with the old API. */
```
Is it useful to mention `(private[ml])` in the JavaDoc? This seems to be duplicated info.

jkbradley replied:
Yeah, I put it there since the Java docs don't handle package private correctly.  Should I remove it?

mengxr replied:
Good point! Let's keep it.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/RandomForestRegressor.scala:
```diff
@@ -0,0 +1,167 @@
+  override def trees: Array[DecisionTreeModel] = _trees.asInstanceOf[Array[DecisionTreeModel]]
+
+  // Note: We may add support for weights (based on tree performance) later on.
+  lazy val _treeWeights: Array[Double] = Array.fill[Double](numTrees)(1.0)
```
private

mengxr commented:
LGTM except some minor inline comments.

jkbradley committed:
small cleanups based on code review

jkbradley commented:
Updated.  The only remaining question is about the ```(private[ml])``` notes.  (See comment above.)

mengxr commented:
test this please

mengxr commented:
test this please

SparkQA commented:
[Test build #698 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/698/consoleFull) for   PR 5626 at commit [`729167a`](https://github.com/apache/spark/commit/729167ab86e406a452509cb4d081efe378d8cacd).

SparkQA commented:
[Test build #698 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/698/consoleFull) for   PR 5626 at commit [`729167a`](https://github.com/apache/spark/commit/729167ab86e406a452509cb4d081efe378d8cacd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Params(`
  * `  case class Params(`
  * `final class GBTClassificationModel(`
  * `trait HasSeed extends Params `
  * `final class GBTRegressionModel(`

 * This patch **adds the following new dependencies:**
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`

 * This patch **removes the following dependencies:**
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`

mengxr commented:
Merged into master. Thanks!

jkbradley commented:
@mengxr  Curious: Why does it say there are unmerged commits?  (I checked, and the last commit was merged correctly.)
